File: Exercices/TGES/csharp_with_dp/src/MailServer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Mail;

namespace TheGreatEmailServer
{
    public class MailServer
    {
        public class Config
        {
            public string storeLocation;
            public List<ValueTuple<string, string>> spamHandlers = new List<(string, string)>();
        }

        private Config config;

        private SpamBlocker spamBlocker;

        private List<MailObserver> observers;

        public MailServer(Config config)
        {
            this.config = config;

            spamBlocker = new SpamBlocker();
            foreach ((string handlerName, string handlerConfig) in config.spamHandlers)
            {
                spamBlocker.AddHandler(SpamHandlerFactory.Create(handlerName, handlerConfig));
            }

            observers = new List<MailObserver>();
        }

        public void Subscribe(MailObserver observer)
        {
            observers.Add(observer);
        }

        public void Unsubscribe(MailObserver observer)
        {
            observers.Remove(observer);
        }

        public void Receive(MailMessage mail)
        {
            NotifyMailReceived(mail);

            if (spamBlocker.ShouldBlock(mail))
            {
                NotifyMailRejected(mail);
                return;
            }

            DispatchMail(mail);
            NotifyMailStored(mail);
        }

        private void DispatchMail(MailMessage mail)
        {
            foreach (var address in mail.To)
            {
                var file = new FileInfo(Path.Combine(config.storeLocation, address.Address, $"{DateTime.Now.ToFileTimeUtc()}.eml"));
                file.Directory.Create();
                File.WriteAllText(file.FullName, mail.ToString()); // Sadly MailMessage has no simple way of converting to string
            }
        }

        private void NotifyMailReceived(MailMessage mail)
        {
            observers.ForEach(observer => observer.MailReceived(mail));
        }

        private void NotifyMailRejected(MailMessage mail)
        {
            observers.ForEach(observer => observer.MailRejected(mail));
        }

        private void NotifyMailStored(MailMessage mail)
        {
            observers.ForEach(observer => observer.MailStored(mail));
        }
    }
}

File: Exercices/TGES/ruby/src/mail_server.rb
class MailServer
  def initialize(config:)
    @config = config

    @observers = []

    @spam_blocker = SpamBlocker.new
    Array(@config.dig('spam_blocker', 'handlers')).each do |name, its_config|
      @spam_blocker.add_handler(SpamHandlers.const_get(name).new(its_config))
    end
  end
  
  def subscribe(observer)
    @observers.push(observer)
  end
  
  def unsubscribe(observer)
    @observers.delete(observer)
  end

  # Called for every incoming mail as a Mail object
  def receive(mail)
    notify_observers(:mail_received, mail)
    
    if @spam_blocker.should_block?(mail)
      notify_observers(:mail_rejected, mail)
      return
    end
    
    dispatch(mail)
    notify_observers(:mail_stored, mail)
  end
  
  protected
  
  def dispatch(mail)
    mail.to.each do |recipient|
      target_dir = File.join(@config["store_location"], recipient)
      Dir.mkdir(target_dir) unless Dir.exist?(target_dir)
      File.write(File.join(target_dir, "#{Time.now.to_i}.eml"), mail.raw_source)
    end
  end
  
  def notify_observers(event, payload)
    @observers.each {|observer| observer.send(event, payload) }
  end
end
